### PR TITLE
apiv2 /containers/json limit differ from docker-api

### PR DESCRIPTION
**Pull request by zhangguanzhang**
Fixes: https://github.com/containers/podman/issues/7722
Signed-off-by: zhangguanzhang <zhangguanzhang@qq.com>

### CODE REVIEW

**Review by edsantiago — CHANGES_REQUESTED** [review]
I still need to review and understand the main aspect of this, but ITM you need to fix the test.

File: test/apiv2/20-containers.at [hunk 1]
```diff
@@ -178,4 +178,24 @@ t GET containers/$cid/json 200 \
...
+t POST  containers/${cid}/start '' 204
+# make sure the container is running
+t GET containers/$cid/json 200 \
+  .State.Status="running"
+
+# 0 means unlimited, need same with docker
+t GET containers/json?limit=0 200 \
+  .[0].Id~[0-9a-f]\\{64\\}
+
+t POST containers/${cid}/stop 204
```
[the remark this answers is not in the document]
Comment (zhangguanzhang) [reply test/apiv2/20-containers.at:198]: thanks

**Comment by edsantiago** [comment]
Friendly amendment? Defer deletion of the previous container; this way we test a multi-container `ps`:
```diff
diff --git a/test/apiv2/20-containers.at b/test/apiv2/20-containers.at
index d635702e4..60ebe2820 100644
--- a/test/apiv2/20-containers.at
+++ b/test/apiv2/20-containers.at
@@ -176,26 +176,27 @@ t GET containers/$cid/json 200 \
   .Config.Cmd='[]' \
   .Path="echo" \
   .Args[0]="param1"
-t DELETE containers/$cid 204

 # create a running container for after
 t POST containers/create '"Image":"'$IMAGE'","Entrypoint":["top"]' 201 \
   .Id~[0-9a-f]\\{64\\}
-cid=$(jq -r '.Id' <<<"$output")
-t GET containers/$cid/json 200 \
+cid_top=$(jq -r '.Id' <<<"$output")
+t GET containers/$cid_top/json 200 \
   .Config.Entrypoint[0]="top" \
   .Config.Cmd='[]' \
   .Path="top"
-t POST  containers/${cid}/start '' 204
+t POST  containers/${cid_top}/start '' 204
 # make sure the container is running
-t GET containers/$cid/json 200 \
+t GET containers/$cid_top/json 200 \
   .State.Status="running"

 # 0 means unlimited, need same with docker
-t GET containers/json?limit=0 200 \
-  .[0].Id~[0-9a-f]\\{64\\}
+t GET 'containers/json?limit=0&all=1' 200 \
+  .[0].Id~[0-9a-f]\\{64\\} \
+  .[1].Id~[0-9a-f]\\{64\\}

-t POST containers/${cid}/stop 204
+t POST containers/${cid_top}/stop '' 204
+t DELETE containers/$cid_top 204
 t DELETE containers/$cid 204

 # vim: filetype=sh
```

**Comment by zhangguanzhang** [comment]
done

**Comment by TomSweeneyRedHat** [comment]
LGTM

File: pkg/api/handlers/compat/containers.go [hunk 2]
```diff
@@ -85,6 +85,10 @@ func ListContainers(w http.ResponseWriter, r *http.Request) {
 		utils.InternalServerError(w, err)
 		return
 	}
+	// 0 means unlimited, need same with docker
+	if query.Limit == 0 {
+		query.Limit = -1
+	}
 	if _, found := r.URL.Query()["limit"]; found && query.Limit != -1 {
```
Comment (edsantiago) [inline pkg/api/handlers/compat/containers.go:92]: Looking this over with more time, would it make more sense to change this line instead, to `&& query.Limit > 0` ?

Comment (zhangguanzhang) [reply pkg/api/handlers/compat/containers.go]: done

**Pull request merged**